File: src/main/java/com/team3/gudit/auth/oauth2/CustomOAuth2User.java
package com.team3.gudit.auth.oauth2;

import com.team3.gudit.user.domain.entity.User;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.jspecify.annotations.Nullable;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.core.user.OAuth2User;

import java.util.Collection;
import java.util.List;
import java.util.Map;

@Getter
@RequiredArgsConstructor
public class CustomOAuth2User implements OAuth2User {
    private final User user;

    private final AuthProvider provider;

    private final OAuth2UserInfo userInfo;

    private final Map<String, Object> attributes;

    private final String nameAttributeKey;

    public static CustomOAuth2User unregistered(AuthProvider provider, OAuth2UserInfo userInfo, Map<String, Object> attributes, String nameAttributeKey) {
        return new CustomOAuth2User(null, provider, userInfo, attributes, nameAttributeKey);
    }

    public boolean isRegistered() {
        return user != null;
    }

    @Override
    public Map<String, Object> getAttributes() {
        return attributes;
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {

        if ( user == null ) {
            return List.of(new SimpleGrantedAuthority("USER"));
        }

        return List.of(new SimpleGrantedAuthority(user.getRole().name()));
    }

    @Override
    public String getName() {
        return String.valueOf(attributes.get(nameAttributeKey));
    }
}

File: src/main/kotlin/com/team3/gudit/auth/oauth2/CustomOAuth2User.kt
package com.team3.gudit.auth.oauth2

import com.team3.gudit.user.domain.entity.User
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.core.user.OAuth2User

class CustomOAuth2User(
    val user: User?,
    val provider: AuthProvider,
    val userInfo: OAuth2UserInfo,
    private val attributes: Map<String, Any>,
    private val principalName: String,
) : OAuth2User {
    fun isRegistered(): Boolean = user != null
    override fun getAttributes(): Map<String, Any> = attributes
    override fun getAuthorities(): Collection<GrantedAuthority> =
        listOf(SimpleGrantedAuthority(user?.role?.name ?: "USER"))
    override fun getName(): String = principalName
}
